Global Styles: Fix incorrect usage of ItemGroup in the Background image panel

im3dabasia opened the pull request:
Part of: https://github.com/WordPress/gutenberg/issues/67425

## What?
Improved the `ItemGroup` component to handle role attributes correctly based on its rendered HTML element. When rendered as a span, it now applies `role="none"` to prevent incorrect semantic meaning, and when rendered as a button, it correctly applies `role="button"`.


## Why?

The previous implementation incorrectly applied `role="list"` in scenarios where a semantic list wasn't appropriate, causing accessibility issues. This change ensures we maintain proper accessibility standards by applying the correct ARIA roles based on the actual HTML element being rendered.


## How?
Updated the component to conditionally set the role attribute based on the as prop:

1. When as="span", it sets `role="none"` to remove any implied semantic meaning
2. Otherwise, it applies `role="button`" for proper accessibility

## Testing Instructions

1. Navigate to Appearance > Editor
2. Navigate to Global Styles > Background.
3. Inspect rendered Button labeled 'Add background image'  and confirm they no longer include `role="list"` when unnecessary.
4. Validate that UI behavior and appearance remain unaffected.



|Before|After|
|-|-|
|<img width="1470" alt="Screenshot 2025-01-13 at 3 04 57 PM" src="https://github.com/user-attachments/assets/9d7be483-5629-4357-a360-9ba89d6c96fe" />|<img width="1470" alt="Screenshot 2025-01-13 at 3 14 12 PM" src="https://github.com/user-attachments/assets/9f7bb698-acfc-4a48-b4b5-e7941fb6a204" />
|

im3dabasia committed:
: a11y issue in background image panel

im3dabasia commented:
Hey @afercia,

Whenever you have a moment, could you please review my PR? It addresses another instance of incorrect usage of the ItemGroup component in the Background Image panel.

I’d really appreciate your feedback!

afercia commented on packages/block-editor/src/components/background-image-control/index.js:
```diff
@@ -125,7 +125,12 @@ function InspectorImagePreviewItem( {
 	return (
-		<ItemGroup as={ as } className={ className } { ...toggleProps }>
+		<ItemGroup
+			role="none"
```
When the background image is set, ItemGroup is rendered as a button element. The role=none would override the native button role, which isn't okay.

im3dabasia replied:
Maybe we can solve this with `imgUrl ? "none" : "button"`. I'll try it out and update you.

afercia replied:
Or maybe based on the `as` prop. However, this would be just a temporary workaround.
This component is still used incorrectly and against its documentation. We can implement a temporary fix but the long term goal is to not allow this kind of incorrect usage.

afercia reviewed (changes requested):
@im3dabasia thanks for your PR. As mentioned in https://github.com/WordPress/gutenberg/issues/67425#issuecomment-2514117966 when the background image is set, this ItemGroup becomes a button element. Setting `role=none` on the button would override the native button role and assistive technology would not communicate this as a button. 
One option could be setting the role conditionally but it wouldn't be great having a button element with role=button.
I'd rather remove the ItemGroup entirely and explore a design change.

im3dabasia commented:
> when the background image is set, this ItemGroup becomes a button element.

Thank you for your feedback @afercia .

I agree that this is happening, but the button has `role="list"`, which I believe is incorrect. While redundant, it would be better to use `role="button"` rather than having a button element with `role="list"`.

<img width="544" alt="image" src="https://github.com/user-attachments/assets/7778aba7-e9f3-4c7e-b5e0-024fdc58fd21" />

As a temporary fix, I’d like to propose two possible solutions. Let me know which one you think works best:


1. Add role prop value conditionally

```jsx
<ItemGroup
  role={ as === 'span' ? 'none' : 'button' }
  as={as}
  className={className}
  {...toggleProps}
/>
```
This would add role="none" by default when a span is rendered and otherwise apply role="button".

3. Add role prop only when `as === 'span'`

```jsx
<ItemGroup
  {...(as === 'span' ? { role: 'none' } : {})}
  as={as}
  className={className}
  {...toggleProps}
/>
```

This approach adds the role prop only when a span is rendered; otherwise, it allows the default behavior to prevail (i.e., a button with role="list").

Personally, I believe the first approach is better as it improves accessibility overall for this issue.


Looking forward to here your thoughts @afercia

afercia commented:
> a button with role="list"

We should avoid that so the first approach is better. However, this can only be acceptable as a temporary solution.
Please add an inline `// TODO: ` comment to explain why there is a temporary fix for the role. It would be good to use the same 'todo' comment requested also for https://github.com/WordPress/gutenberg/pull/68633

im3dabasia committed:
fix: Role of button when rendered in span/button

im3dabasia commented:
Hi @afercia,

Sorry for the delay. I have updated the PR with the relevant changes as discussed. I have also added a TODO message for future reference. Please review my PR once again and provide any feedback.

Thank you for your time. I appreciate it.

im3dabasia committed:
revert: Fix unwanted storybook change

t-hamano reviewed (commented):
We need to explore a temporary fix here, but I don't think simply changing the `role` attribute is enough.

Here are what I consider to be the problems.

The focus outline style is slightly different than that of the `Button` component:

![image](https://github.com/user-attachments/assets/d6f58d69-b688-44c7-b9fe-532c3edd55cb)

The focus outline color is black:

![image](https://github.com/user-attachments/assets/931b640f-5154-41f2-8edb-1b9c52d2cbe8)

These issues seem to be caused by customizing the `ItemGroup` component.

I would suggest, similar to #68633, to remove the `ItemGroup` component completely and use the `Button` component. We'll probably need to adjust the implementation of the `InspectorImagePreviewItem` component. What do you think?

im3dabasia committed:
fix: Replace ItemGroup with Button

im3dabasia commented:
> I would suggest, similar to #68633, to remove the `ItemGroup` component completely and use the `Button` component. We'll probably need to adjust the implementation of the `InspectorImagePreviewItem` component. What do you think?


Correct! I have adjusted the code in `InspectorImagePreviewItem`.

I have replaced `ItemGroup` with `Button` and am rendering the button conditionally when we have an Image URL. This prevents the issue of having 2 buttons piled up when no image exists (which happens when we initially come to the page, before uploading any image - as shown in the screenshot below).

<img width="390" alt="Screenshot 2025-03-27 at 3 02 54 PM" src="https://github.com/user-attachments/assets/37290f98-ea6f-4469-b965-6632babf4adf" />

This approach solves our problem and also eliminates the `ItemGroup` usage. Let me know what you think!

im3dabasia commented:
@t-hamano , 

When you get a chance. Please review this PR. I feel we are close to shipping this 🚢

t-hamano reviewed (commented):
Thanks for the update! 

Can you fix the button when the background image panel is open? It seems that the button elements are nested:

![image](https://github.com/user-attachments/assets/c042133c-2852-4984-9f07-28f73e59a7b3)

Additionally, we might be able to extract `FlexItem`:

```jsx
function InspectorImagePreviewItem( {} ) {
	// ...
	const renderLabelContent = () => (
		<FlexItem>
			<Truncate>{ label }</Truncate>
			{/* ... */}
		</FlexItem>
	);

	return imgUrl ? (
		<Button>
			<HStack className="block-editor-global-styles-background-panel__inspector-preview-inner">
				{ imgUrl && {/* ... */} }
				{ renderLabelContent() }
			</HStack>
		</Button>
	) : (
		renderLabelContent()
	);
}
```

im3dabasia committed:
fix: Breading button nesting in popover

im3dabasia committed:
chore: Change render function name

t-hamano reviewed (commented):
@im3dabasia Can you investigate and fix the browser console error occurring when adding a background image? This error doesn't occur on trunk:


https://github.com/user-attachments/assets/604b0ae4-c79a-4c2f-a7e5-3f79acd97e02

im3dabasia committed:
fix: Error related to isOpen prop in button

im3dabasia commented:
@t-hamano , The console error was occurring because the `isOpen` prop was passed to the Button component and was directly passed onto the button element. I have fixed this issue.

Please check once again.

t-hamano reviewed (approved):
LGTM!

Finally, can we do a small refactor like this?

```jsx
const { isOpen, ...restToggleProps } = toggleProps;

useEffect( () => {
	if ( typeof isOpen !== 'undefined' ) {
		onToggleCallback( isOpen );
	}
}, [ isOpen, onToggleCallback ] );
```

im3dabasia committed:
chore: Code refactor

im3dabasia commented:
@t-hamano . Thanks for the continuous iterations of feedback. Please check when you have a moment
